packages/tabsheet/src/vaadin-tabsheet.js

```diff
@@ -3,7 +3,269 @@
+ * @fires {CustomEvent} selected-changed - Fired when the `selected` property changes.
+ *
+ * @extends HTMLElement
+ * @mixes ElementMixin
+ * @mixes ThemableMixin
+ * @mixes ControllerMixin
+ */
+class TabSheet extends ControllerMixin(ElementMixin(ThemableMixin(PolymerElement))) {
+  static get template() {
+    return html`
+      <style>
+        :host([hidden]) {
+          display: none !important;
+        }
+
+        :host {
+          display: flex;
+          height: 400px;
+        }
+
+        :host([orientation='horizontal']) {
+          flex-direction: column;
+        }
+
+        [part='tabs-container'] {
+          display: flex;
+          flex-direction: column;
+          align-items: baseline;
+        }
+
+        :host([orientation='horizontal']) [part='tabs-container'] {
+          flex-direction: row;
+        }
+
+        [part='panels'] {
+          overflow: auto;
+          flex: 1;
+        }
+      </style>
+
+      <div part="tabs-container">
+        <slot name="prefix"></slot>
+        <slot name="tabs"></slot>
+        <slot name="suffix"></slot>
+      </div>
+
+      <div part="panels">
+        <slot id="panel-slot"></slot>
+      </div>
+    `;
+  }
+
+  static get is() {
+    return 'vaadin-tabsheet';
+  }
+
+  static get properties() {
+    return {
+      /**
+       * The tabsheet's orientation. Possible values are `horizontal|vertical`
+       * @type {!TabSheetOrientation}
+       */
+      orientation: {
+        reflectToAttribute: true,
+        value: 'horizontal',
+        type: String,
+        observer: '__orientationChanged',
+      },
+
+      /**
+       * The list of `<vaadin-tab>`s from which a selection can be made.
+       * It is populated from the elements passed inside the slotted
+       * `<vaadin-tabs>`, and updated dynamically when adding or removing items.
+       *
+       * Note: unlike `<vaadin-combo-box>`, this property is read-only.
+       * @type {!Array<!Tab> | undefined}
+       */
+      items: {
+        type: Array,
+        readOnly: true,
+        notify: true,
+      },
+
+      /**
+       * The index of the selected tab.
+       */
+      selected: {
+        value: 0,
+        type: Number,
+        reflectToAttribute: true,
+        notify: true,
+      },
+
+      /**
+       * The slotted <vaadin-tabs> element.
+       */
+      __tabs: {
+        type: Object,
+      },
+
+      /**
+       * The panel elements.
+       */
+      __panels: {
+        type: Array,
+      },
+    };
+  }
+
+  /** @protected */
+  ready() {
+    super.ready();
+
+    this.role = 'tablist';
+
+    const tabsItemsChangedListener = () => this._setItems(this.__tabs.items);
+
+    const tabsSelectedChangedListener = () => {
+      this.selected = this.__tabs.selected;
+    };
+
+    // Observe the tabs slot for a <vaadin-tabs> element.
+    this.addController(
+      new (class extends SlotController {
```

Comment:
nit: I would prefer to move the controller class definition outside the method call.

Reply:
Done